### PR TITLE
Default Result Assertions Fix (again)

### PR DESCRIPTION
**Pull request by JordanllHarper**
______________________________________
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/main/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: apis/fluent-en_GB/extensions/atrium-api-fluent-en_GB-kotlin_1_3/src/commonTest/kotlin/ch/tutteli/atrium/api/fluent/en_GB/kotlin_1_3/samples/ResultExpectationSamples.kt [hunk 1]
```diff
@@ -57,7 +57,22 @@ class ResultExpectationSamples {
...
+//
+//
+//
+//        expect(failure) {
+//            toBeAFailure<IllegalArgumentException>()
+//        }
+//
+//
+//    }
+//
```
Comment (robstoll) [inline apis/fluent-en_GB/extensions/atrium-api-fluent-en_GB-kotlin_1_3/src/commonTest/kotlin/ch/tutteli/atrium/api/fluent/en_GB/kotlin_1_3/samples/ResultExpectationSamples.kt:75]: remove this

File: apis/fluent-en_GB/extensions/atrium-api-fluent-en_GB-kotlin_1_3/src/commonTest/kotlin/ch/tutteli/atrium/api/fluent/en_GB/kotlin_1_3/samples/ResultExpectationSamples.kt [hunk 2]
```diff
@@ -57,7 +57,22 @@ class ResultExpectationSamples {
...
 //    @Test
 //    fun toBeAFailureFeature() {
 //        val message = "wrong argument"
```
Comment (robstoll) [inline apis/fluent-en_GB/extensions/atrium-api-fluent-en_GB-kotlin_1_3/src/commonTest/kotlin/ch/tutteli/atrium/api/fluent/en_GB/kotlin_1_3/samples/ResultExpectationSamples.kt:78]: uncomment all the tests instead

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/ResultExpectationsSpec.kt [hunk 3]
```diff
@@ -44,16 +44,16 @@ abstract class ResultExpectationsSpec(
...
-//    include(object : AssertionCreatorSpec<Result<Int>>(
-//        "$describePrefix[failure] ", Result.failure(IllegalArgumentException("oh no...")),
-//        assertionCreatorSpecTriple(
-//            toBeAFailure.name,
-//            "${DescriptionCharSequenceExpectation.VALUE.getDefault()}: \"oh no...\"",
-//            { apply { toBeAFailure.invoke(this) { messageToContain("oh no...") } } },
-//            { apply { toBeAFailure.invoke(this) {} } }
-//        )
-//    ) {})
+    //TODO 0.20.0 activate once we have the workaround for #1234 implemented
```
Comment (robstoll) [inline misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/ResultExpectationsSpec.kt:47]: remove todo

File: misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/ResultExpectationsSpec.kt [hunk 4]
```diff
@@ -120,22 +120,22 @@ abstract class ResultExpectationsSpec(
...
-//                        expect(resultFailure).toBeAFailureFun { messageToContain("oh yes...") }
-//                    }.toThrow<AssertionError> {
-//                        messageToContain(
-//                            "$exceptionDescr: ${IllegalArgumentException::class.fullName}",
-//                            DescriptionCharSequenceExpectation.TO_CONTAIN.getDefault(), "${DescriptionCharSequenceExpectation.VALUE.getDefault()}: \"oh yes...\""
-//                        )
-//                    }
-//                }
-//            }
+            //TODO 0.20.0 activate once we have the workaround for #1234 implemented
```
Comment (robstoll) [inline misc/atrium-specs/src/commonMain/kotlin/ch/tutteli/atrium/specs/integration/ResultExpectationsSpec.kt:123]: remove todo

**Review by robstoll — CHANGES_REQUESTED** [review]

**Review by robstoll — APPROVED** [review]

File: logic/extensions/atrium-logic-kotlin_1_3/src/commonMain/kotlin/ch/tutteli/atrium/logic/kotlin_1_3/impl/DefaultResultAssertions.kt [hunk 5]
```diff
@@ -33,8 +36,42 @@ class DefaultResultAssertions : ResultAssertions {
...
+
+
+            //fix is here for bug in kotlin 1.3, 1.4, 1.5 (fixed in 1.6) => https://youtrack.jetbrains.com/issue/KT-50974
+            //Comments below explain the fix
+            val badResult = exceptionOrNull()
+            //mapping manually to get internal value throwable - exception or null just produced null + we lost the exception
+            val maybeSubjectResult = container.maybeSubject.map {
+                onSuccess { null }
+                onFailure { it.cause }
+            }
```
Comment (robstoll) [inline logic/extensions/atrium-logic-kotlin_1_3/src/commonMain/kotlin/ch/tutteli/atrium/logic/kotlin_1_3/impl/DefaultResultAssertions.kt:50]: sorry, I did not really find the time beforehand to take a proper look. This code does nothign as your onSuccess and onFailure only perform side effects. They don't return something which means, maybeSubjectResult is the same as `container.maybeSubject`. You don't have to do anything. I will improve the code and let you know my result in the end

Comment (JordanllHarper) [reply logic/extensions/atrium-logic-kotlin_1_3/src/commonMain/kotlin/ch/tutteli/atrium/logic/kotlin_1_3/impl/DefaultResultAssertions.kt:50]: Ah I see the error I made sorry about that. Would the solution to be to return some sort of value from the success or failure states and check that?

Comment (robstoll) [reply logic/extensions/atrium-logic-kotlin_1_3/src/commonMain/kotlin/ch/tutteli/atrium/logic/kotlin_1_3/impl/DefaultResultAssertions.kt:50]: the statement doesn't do anything and you still have fixed the issue which means these lines we can just delete :)

**Pull request merged**